File: CPP/ByRound/2036/2036_D_I_Love_1543.cpp
// Problem: CF 2036 D - I Love 1543
// https://codeforces.com/contest/2036/problem/D

/*
D. I Love 1543
Purpose: Count occurrences of the sequence "1543" in each concentric layer of a grid when traversed clockwise.

Algorithms/Techniques:
- Layer-by-layer traversal of the grid in clockwise direction.
- For each layer, collect characters in order and check for pattern "1543".
- Uses vector to store layer characters and string for pattern matching.

Time Complexity: O(n * m) per test case, where n and m are dimensions of the grid.
Space Complexity: O(n * m) for storing the grid and temporary vectors for each layer.

*/

#include <cstdio>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;
int t, n, m, i, j, k, l, x, y, c;
char s[1000005], a[1005][1005];
int main() {
  scanf("%d", &t);
  while (t--) {
    scanf("%d%d", &n, &m);
    for (i = 0; i < n; i++)
      scanf("%s", a[i]);
    c = 0;
    for (l = 0; l < min(n, m) / 2; l++) {
      vector<char> v;
      int u = n - 1 - l, w = m - 1 - l;
      // Traverse top row left to right
      for (j = l; j <= w; j++)
        v.push_back(a[l][j]);
      // Traverse right column top to bottom
      for (i = l + 1; i <= u; i++)
        v.push_back(a[i][w]);
      // Traverse bottom row right to left (if not a single row)
      if (u > l)
        for (j = w - 1; j >= l; j--)
          v.push_back(a[u][j]);
      // Traverse left column bottom to top (if not a single column)
      if (w > l)
        for (i = u - 1; i > l; i--)
          v.push_back(a[i][l]);
      int sz = v.size();
      string s(v.begin(), v.end());
      // Check for pattern "1543" in the layer
      for (k = 0; k < sz; k++) {
        if (s[k] == '1' && s[(k + 1) % sz] == '5' && s[(k + 2) % sz] == '4' &&
            s[(k + 3) % sz] == '3')
          c++;
      }
    }
    printf("%d\n", c);
  }
}


// https://github.com/VaHiX/codeForces/

File: Python/ByRound/2036/2036_D_I_Love_1543.py
# Problem: CF 2036 D - I Love 1543
# https://codeforces.com/contest/2036/problem/D

"""
D. I Love 1543
Algorithm: Spiral Traversal + Pattern Matching
Time Complexity: O(n * m) per test case, where n and m are dimensions of the carpet.
Space Complexity: O(n * m) for storing the input grid and intermediate string representations.

The problem requires counting how many times the pattern "1543" appears in the spiral traversal 
of layers of a grid. Each layer is traversed clockwise, and multiple occurrences (overlapping 
or non-overlapping) are counted. 

For each layer:
- Extract top row (left to right)
- Extract right column (top to bottom, skipping corners)
- Extract bottom row (right to left)
- Extract left column (bottom to top, skipping corners)
Concatenate these four parts into one string.
Then, we append the first 3 characters of this string to itself and count how many times 
"1543" appears in it — to handle wraparound patterns that span across layer boundaries.

This approach processes each cell once per layer it belongs to.
"""

for _ in range(int(input())):
    n, m = map(int, input().split())
    g = [input() for _ in range(n)]
    t = 0
    L = min(n, m) // 2  # Number of layers to process
    for i in range(L):
        # Extract top row from left to right
        p1 = g[i][i : m - i]
        # Extract right column from top to bottom (excluding corners)
        p2 = "".join(g[j][m - i - 1] for j in range(i + 1, n - i - 1))
        # Extract bottom row from right to left
        p3 = g[n - i - 1][i : m - i][::-1]
        # Extract left column from bottom to top (excluding corners)
        p4 = "".join(g[j][i] for j in range(i + 1, n - i - 1))[::-1]
        # Concatenate all four parts to form the layer's spiral string
        s = p1 + p2 + p3 + p4
        # Append first 3 characters to handle wraparound cases (e.g., last few chars 
        # of current layer may match with start of next)
        t += (s + s[:3]).count("1543")
    print(t)


# https://github.com/VaHiX/codeForces/